src/core/resources/positions/index.ts

```diff
@@ -0,0 +1,315 @@
+import {
+  QueryConfig,
+  QueryFunctionArgs,
+  QueryFunctionResult,
+  createQueryKey,
+  queryClient,
+} from '~/core/react-query';
+import { SupportedCurrencyKey } from '~/core/references';
+import { AssetApiResponse, ParsedUserAsset } from '~/core/types/assets';
+import { ChainId, ChainName } from '~/core/types/chains';
+import { parseUserAsset } from '~/core/utils/assets';
+import {
+  chainIdFromChainName,
+  getBackendSupportedChains,
+} from '~/core/utils/chains';
+import { RainbowError, logger } from '~/logger';
+
+const POSITIONS_TIMEOUT_DURATION = 20000;
+
+// ///////////////////////////////////////////////
+// Response Types
+export type AddysPositionAsset = Omit<
+  AssetApiResponse,
+  'type' | 'interface'
+> & {
+  verified: boolean;
+};
+
+export interface AddysPosition {
+  borrows: { asset: AddysPositionAsset; quantity: string }[];
+  claimables: { asset: AddysPositionAsset; quantity: string }[];
+  dapp: {
+    colors: { primary: string; fallback: string; shadow: string };
+    icon_url: string;
+    name: string;
+    url: string;
+  };
+  deposits: { asset: AddysPositionAsset; quantity: string }[];
+  network: ChainName;
+  stakes: { asset: AddysPositionAsset; quantity: string }[];
+  type: string;
+}
+
+export type ParsedAddysPosition = Omit<
+  AddysPosition,
+  'borrows' | 'claimables' | 'deposits' | 'stakes'
+> & {
+  borrows: ParsedUserAsset[];
+  claimables: ParsedUserAsset[];
+  deposits: ParsedUserAsset[];
+  stakes: ParsedUserAsset[];
+  chainId: ChainId;
+};
+
+export interface AddysPositionsResponse {
+  meta: {
+    addresses: Address[];
+    addresses_with_errors: Address[];
+    chain_ids: number[];
+    chain_ids_with_errors: number[];
+    currency: string;
+    errors: string[];
+    status: string | number;
+  };
+  payload: {
+    positions: AddysPosition[];
+  };
+  status: number;
+}
+
+export type ParsedPositionsByChain = Record<
+  ChainId,
+  {
+    positions: ParsedAddysPosition[];
+  }
+>;
+
+// ///////////////////////////////////////////////
+// Query Types
+
+export type PositionsArgs = {
+  address: Address;
+  currency: SupportedCurrencyKey;
+  testnetMode: boolean;
+};
+
+// ///////////////////////////////////////////////
+// Query Key
+
+const positionsQueryKey = ({ address, currency, testnetMode }: PositionsArgs) =>
+  createQueryKey(
+    'positions',
+    { address, currency, testnetMode },
+    { persisterVersion: 1 },
+  );
+
+type PositionsQueryKey = ReturnType<typeof positionsQueryKey>;
+
+// ///////////////////////////////////////////////
+// Query Function
+
+async function positionsQueryFunction({
+  queryKey: [{ address, currency, testnetMode }],
+}: QueryFunctionArgs<typeof positionsQueryKey>) {
+  if (!address) return {} as ParsedPositionsByChain;
+  const cache = queryClient.getQueryCache();
+  const cachedPositions = (cache.find(
+    positionsQueryKey({ address, currency, testnetMode }),
+  )?.state.data || {}) as ParsedPositionsByChain;
```

Comment:
have you tried without this? I have a feeling react query keeps the already cache data if the next refetch fails
like the query status is error but it may still have data (?)

Reply:
what if the BE reports an error with a particular chain id? That won't throw, but then we'll populate an empty set of positions because this object contains positions for all chains.